src/cloudWatchLogs/timeFilterSubmenu.ts

```diff
@@ -0,0 +1,171 @@
+/*!
+ * Copyright 2020 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { Prompter, PromptResult } from '../shared/ui/prompter'
+import { InputBoxPrompter } from '../shared/ui/inputPrompter'
+import { ItemLoadTypes, QuickPickPrompter, DataQuickPickItem, createQuickPick } from '../shared/ui/pickerPrompter'
+import { createInputBox } from '../shared/ui/inputPrompter'
+import { isValidResponse, StepEstimator } from '../shared/wizards/wizard'
+
+export function isViewAllEvents(response: TimeFilterResponse) {
+    return response.start === response.end
+}
+
+export interface TimeFilterResponse {
+    // # of miliseconds since january 1, 1970 since thats what API expects.
+    readonly start: number
+    readonly end: number
+}
+
+const customRange = Symbol('customRange')
+
+export class TimeFilterSubmenu extends Prompter<TimeFilterResponse> {
+    // TODO: Generalize submenu code between this and the region Submenu.
+    private currentState: 'custom-range' | 'recent-range' = 'recent-range'
+    private steps?: [current: number, total: number]
+    public defaultPrompter: QuickPickPrompter<typeof customRange | number> = this.createMenuPrompter()
+    public customPrompter: InputBoxPrompter = this.createDateBox()
+
+    public constructor() {
+        super()
+    }
+
+    private get recentTimeOptions(): ItemLoadTypes<number> {
+        const options: DataQuickPickItem<number>[] = []
+        options.push({
+            label: 'View all events',
+            data: 0,
+        })
+        options.push({
+            label: 'Last 1 Minute',
+            data: 1,
+        })
+        options.push({
+            label: 'Last 30 Minutes',
+            data: 30,
+        })
+        options.push({
+            label: 'Last 1 Hour',
+            data: 60,
+        })
+        options.push({
+            label: 'Last 12 Hours',
+            data: 60 * 12,
+        })
+        return options
+    }
+
+    public createMenuPrompter() {
+        const prompter = createQuickPick<number | typeof customRange>(this.recentTimeOptions)
+
+        prompter.quickPick.items = [
+            ...prompter.quickPick.items,
+            {
+                label: 'Custom time range',
+                data: customRange,
+                detail: `YYYY/MM/DD-YYYY/MM/DD`,
+            },
+        ]
+
+        return prompter
+    }
+
+    private switchState(newState: 'custom-range' | 'recent-range') {
+        this.currentState = newState
+    }
+
+    public createDateBox(): InputBoxPrompter {
+        return createInputBox({
+            title: 'Enter custom date range',
+            placeholder: 'YYYY/MM/DD-YYYY/MM/DD',
+            validateInput: input => this.validateDate(input),
+        })
+    }
+
+    protected async promptUser(): Promise<PromptResult<TimeFilterResponse>> {
+        while (true) {
+            switch (this.currentState) {
```

Comment:
Looks like `currentState` has known expects fields, so the `default` case could throw an error to guard against future changes (and an infinite loop here).

Also I wonder if changing the `while(true)` to something like `for(int i = 0; i < 5; i++)` makes sense, to guard against infinite loop?